packages/grid/x-data-grid/src/hooks/features/columns/gridColumnsUtils.ts

```diff
@@ -339,46 +340,67 @@ export const createColumnsState = ({
   > & {
     lookup: { [field: string]: Omit<GridStateColDef, 'computedWidth'> };
   };
-  if (isInsideStateInitializer || keepOnlyColumnsToUpsert) {
+  if (isInsideStateInitializer) {
     columnsStateWithoutColumnVisibilityModel = {
       all: [],
       lookup: {},
     };
   } else {
     const currentState = gridColumnsSelector(apiRef.current.state);
     columnsStateWithoutColumnVisibilityModel = {
-      all: [...currentState.all],
-      lookup: { ...currentState.lookup },
+      all: keepOnlyColumnsToUpsert ? [] : [...currentState.all],
+      lookup: { ...currentState.lookup }, // Will be cleaned later if keepOnlyColumnsToUpsert=true
     };
   }
 
+  let columnsToKeep: Record<string, boolean> = {};
+  if (keepOnlyColumnsToUpsert && !isInsideStateInitializer) {
+    columnsToKeep = Object.keys(columnsStateWithoutColumnVisibilityModel.lookup).reduce(
+      (acc, key) => ({ ...acc, [key]: false }),
+      {},
+    );
+  }
+
   const columnsToUpsertLookup: Record<string, true> = {};
   columnsToUpsert.forEach((newColumn) => {
-    columnsToUpsertLookup[newColumn.field] = true;
-    if (columnsStateWithoutColumnVisibilityModel.lookup[newColumn.field] == null) {
+    const { field } = newColumn;
+    columnsToUpsertLookup[field] = true;
+    columnsToKeep[field] = true;
+    let existingState = columnsStateWithoutColumnVisibilityModel.lookup[field];
+
+    if (existingState == null) {
       // New Column
-      columnsStateWithoutColumnVisibilityModel.lookup[newColumn.field] = {
+      existingState = {
         ...getGridColDef(columnsTypes, newColumn.type), // TODO v6: Inline `getGridColDef`
-        ...newColumn,
-      };
-      columnsStateWithoutColumnVisibilityModel.all.push(newColumn.field);
-    } else {
-      const mergedColumn = {
-        ...columnsStateWithoutColumnVisibilityModel.lookup[newColumn.field],
-        ...newColumn,
+        field,
+        hasBeenResized: false,
       };
+      columnsStateWithoutColumnVisibilityModel.all.push(field);
+    } else if (keepOnlyColumnsToUpsert) {
+      columnsStateWithoutColumnVisibilityModel.all.push(field);
+    }
 
-      if (
-        !mergedColumn.hasBeenResized &&
-        COLUMNS_DIMENSION_PROPERTIES.some((propertyName) => newColumn[propertyName] !== undefined)
-      ) {
-        mergedColumn.hasBeenResized = true;
-      }
+    const validDimensions = COLUMNS_DIMENSION_PROPERTIES.reduce<string[]>((acc, key) => {
+      const value = newColumn[key];
+      return typeof value === 'number' && value !== Infinity ? [...acc, key] : acc;
+    }, []);
```

Comment:
why not keep the `COLUMNS_DIMENSION_PROPERTIES.some(...)` since you do not reuse `validDimensions` except to test its length?

Reply:
You can do the same with `some` such that we understand directly why there is a loop on `COLUMNS_DIMENSION_PROPERTIES`
```js
const hasValidDimension = COLUMNS_DIMENSION_PROPERTIES.some((key) => {
      const value = newColumn[key];
      return typeof value === 'number' && value !== Infinity;
    }); 
```

Reply:
Updated.